core/services/workflows/execution_strategy.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+
+	"github.com/smartcontractkit/libocr/permutation"
+
+	"golang.org/x/crypto/sha3"
+)
+
+type executionStrategy interface {
+	Apply(ctx context.Context, l logger.Logger, cap capabilities.CallbackCapability, req capabilities.CapabilityRequest) (values.Value, error)
+}
+
+var _ executionStrategy = immediateExecution{}
+
+type immediateExecution struct{}
+
+func (i immediateExecution) Apply(ctx context.Context, lggr logger.Logger, cap capabilities.CallbackCapability, req capabilities.CapabilityRequest) (values.Value, error) {
+	l, err := capabilities.ExecuteSync(ctx, cap, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// `ExecuteSync` returns a `values.List` even if there was
+	// just one return value. If that is the case, let's unwrap the
+	// single value to make it easier to use in -- for example -- variable interpolation.
+	if len(l.Underlying) > 1 {
+		return l, nil
+	}
+
+	return l.Underlying[0], nil
+}
+
+var _ executionStrategy = scheduledExecution{}
+
+type scheduledExecution struct {
+	DON      *capabilities.DON
+	PeerID   *p2ptypes.PeerID
+	Position int
+}
+
+var (
+	// S = [N]
+	Schedule_AllAtOnce = "allAtOnce"
+	// S = [1 * N]
+	Schedule_OneAtATime = "oneAtATime"
+)
+
+// scheduledExecution generates a pseudo-random transmission schedule,
+// and delays execution until a node is required to transmit.
+func (d scheduledExecution) Apply(ctx context.Context, lggr logger.Logger, cap capabilities.CallbackCapability, req capabilities.CapabilityRequest) (values.Value, error) {
+	tc, err := d.transmissionConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := cap.Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	// Case 1: Local DON
+	case info.DON == nil:
+		n := len(d.DON.Members)
+		key := d.key(d.DON.Config.SharedSecret, req.Metadata.WorkflowID, req.Metadata.WorkflowExecutionID)
+		sched, err := schedule(tc.Schedule, n)
+		if err != nil {
+			return nil, err
+		}
+
+		picked := permutation.Permutation(n, key)
+		delay := d.delayFor(d.Position, sched, picked, tc.DeltaStage)
+		if delay == nil {
+			lggr.Debugw("skipping transmission: node is not included in schedule")
+			return nil, nil
+		}
+
+		lggr.Debugf("execution delayed by %+v", *delay)
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(*delay):
+			lggr.Debugw("executing delayed execution")
+			return immediateExecution{}.Apply(ctx, lggr, cap, req)
+		}
+	// Case 2: Remote DON
+	default:
+		// TODO: fill in the remote DON case once consensus has been reach on what to do.
+		lggr.Debugw("remote DON transmission not implemented: using immediate execution")
+		return immediateExecution{}.Apply(ctx, lggr, cap, req)
+	}
+}
+
+func (d scheduledExecution) key(sharedSecret [16]byte, workflowID, workflowExecutionID string) [16]byte {
```

Comment:
is it important across software versions that the return value is constant given the same inputs? asked differently do all members in the don always need to calculate the same permutation?

Reply:
Yes they do, hence the use of a common secret salted with the workflow and workflow execution ID